### PR TITLE
feat(paywall): Add provider

### PR DESCRIPTION
**Pull request by searchableguy**
# Description

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: packages/paywall/example/src/profile.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+export function Profile() {
+  const { address, isConnected } = useAccount()
+  const { connect } = useConnect({
+    connector: new InjectedConnector({
+      options: {
+        name: 'Unlock Paywall Provider',
+        getProvider: () => {
+          const provider = paywall.getProvider('http://localhost:3000')
+          paywall.connect(provider)
```
Comment (julien51) [inline packages/paywall/example/src/profile.tsx:15]: That's not rigth. The call to `connect` is internal to `Paywall` and will be performed (indirectly) when wagmi's `connect` is called (it calls eth_requestAccounts, which the Paywall "intercepts" to call its own internal `connect`).

File: packages/paywall/src/Paywall.ts [hunk 2]
```diff
@@ -96,7 +97,7 @@ export class Paywall {
    * @param unlockUrl
    * @returns
    */
-  getProvider = (unlockUrl?: string, config?: any) => {
+  getProvider = (config: OauthConfig, unlockUrl?: string) => {
```
Comment (julien51) [inline packages/paywall/src/Paywall.ts:100]: I don't thinl we need the OauthConfig here

**Comment by julien51** [comment]
Please move the example to the examples repo https://github.com/unlock-protocol/examples

File: packages/paywall/package.json [hunk 3]
```diff
@@ -21,7 +21,8 @@
     "lint": "eslint --resolve-plugins-relative-to ../eslint-config --ext .tsx,.ts,.js src/",
     "test": "NODE_ENV=test vitest run --coverage",
     "clean": "rm -rf build dist coverage",
-    "ci": "yarn test && yarn build"
+    "ci": "yarn test && yarn build",
+    "example": "vite --config example/vite.config.js example/src"
```
Comment (julien51) [inline packages/paywall/package.json:25]: Move to examples repo

File: packages/paywall/package.json [hunk 4]
```diff
@@ -33,13 +34,20 @@
     "@unlock-protocol/networks": "workspace:^",
     "@unlock-protocol/tsconfig": "workspace:^",
     "@unlock-protocol/types": "workspace:^",
+    "@unlock-protocol/unlock-js": "workspace:^",
+    "@vitejs/plugin-react": "4.0.0",
```
Comment (julien51) [inline packages/paywall/package.json:38]: Move to examples repo

File: packages/paywall/package.json [hunk 5]
```diff
@@ -33,13 +34,20 @@
...
+    "axios": "1.4.0",
     "eslint": "8.41.0",
+    "ethers": "^5.7.0",
```
Comment (julien51) [inline packages/paywall/package.json:41]: Please pin dependencies

File: packages/paywall/package.json [hunk 6]
```diff
@@ -33,13 +34,20 @@
...
     "tsup": "6.7.0",
     "typescript": "5.1.3",
+    "viem": "^1.0.7",
```
Comment (julien51) [inline packages/paywall/package.json:44]: remove

File: packages/paywall/package.json [hunk 7]
```diff
@@ -33,13 +34,20 @@
...
     "vite": "4.3.9",
     "vite-plugin-css-injected-by-js": "3.1.1",
+    "vite-plugin-node-polyfills": "0.9.0",
```
Comment (julien51) [inline packages/paywall/package.json:47]: needed?

File: packages/paywall/package.json [hunk 8]
```diff
@@ -33,13 +34,20 @@
...
     "vitest": "0.31.4",
-    "vitest-fetch-mock": "0.2.2"
+    "vitest-fetch-mock": "0.2.2",
+    "wagmi": "1.2.0"
```
Comment (julien51) [inline packages/paywall/package.json:50]: remove

File: packages/paywall/src/Paywall.ts [hunk 9]
```diff
@@ -5,15 +5,16 @@ import { dispatchEvent, unlockEvents, injectProviderInfo } from './utils'
...
+  getInjectedProvider,
+  enableProvider,
+  PaywallProvider,
+} from './utils/provider'
 import { unlockAppUrl } from './urls'
 
 export const checkoutIframeClassName = 'unlock-protocol-checkout'
 
+Postmate.debug = true
+
```
Comment (julien51) [inline packages/paywall/src/Paywall.ts:17]:
```suggestion
```

**Review by julien51 — COMMENTED** [review]
We need to bump the version of the npm module

File: packages/paywall/src/utils/provider.ts [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+          callback(args.error, args.response)
+        }
+      }
+    )
+  }
+
+  async #createRequestPromise(args: MethodCall) {
+    return new Promise((resolve, reject) => {
+      this.#methodCalls[args.id] = (error: any, response: any) => {
+        console.log('Got response!')
```
Comment (julien51) [inline packages/paywall/src/utils/provider.ts:111]: Remove!

File: packages/paywall/package.json (hunk 3 above)
Comment (searchableguy) [reply packages/paywall/package.json:25]: I don't think we should move this. It is good to have it in the same folder so can quickly test things for any changes we make.

File: packages/paywall/tsconfig.json [hunk 11]
```diff
@@ -6,7 +6,8 @@
     ],
     "outDir": "build",
     "moduleResolution": "node",
-    "noImplicitAny": false
+    "noImplicitAny": false,
+    "jsx": "react-jsx"
```
Comment (julien51) [inline packages/paywall/tsconfig.json:10]: remove!

File: unlock-app/package.json [hunk 12]
```diff
@@ -22,6 +22,7 @@
     "@unlock-protocol/crypto-icon": "workspace:./packages/crypto-icon",
     "@unlock-protocol/eslint-config": "workspace:./packages/eslint-config",
     "@unlock-protocol/networks": "workspace:./packages/networks",
+    "@unlock-protocol/paywall": "workspace:./packages/paywall",
```
Comment (julien51) [inline unlock-app/package.json:25]: Any reason for this?

File: unlock-app/src/hooks/useCheckoutCommunication.ts [hunk 13]
```diff
@@ -116,23 +120,65 @@ export const useCheckoutCommunication = () => {
...
       setPaywallConfig(config)
     },
-    authenticate: () => {
+    authenticate: (config: any) => {
       setOauthConfig({
-        clientId: window.parent.location.host,
-        responseType: '',
-        state: '',
-        redirectUri: '',
+        clientId: config?.clientId ?? 'http://localhost:3000',
```
Comment (julien51) [inline unlock-app/src/hooks/useCheckoutCommunication.ts:135]:
```suggestion
        clientId: config?.clientId,
```

File: unlock-app/src/hooks/usePostmateParent.ts [hunk 14]
```diff
@@ -1,6 +1,8 @@
 import { useState, useEffect } from 'react'
 import Postmate from 'postmate'
 
+Postmate.debug = true
+
```
Comment (julien51) [inline unlock-app/src/hooks/usePostmateParent.ts:5]:
```suggestion
```

File: packages/paywall/package.json [hunk 15]
```diff
@@ -33,15 +33,18 @@
     "@unlock-protocol/networks": "workspace:^",
     "@unlock-protocol/tsconfig": "workspace:^",
     "@unlock-protocol/types": "workspace:^",
+    "@unlock-protocol/unlock-js": "workspace:^",
+    "@vitejs/plugin-react": "4.0.0",
```
Comment (searchableguy) [inline packages/paywall/package.json:37]:
```suggestion
```

File: packages/paywall/package.json [hunk 16]
```diff
@@ -33,15 +33,17 @@
     "@unlock-protocol/networks": "workspace:^",
     "@unlock-protocol/tsconfig": "workspace:^",
     "@unlock-protocol/types": "workspace:^",
+    "@unlock-protocol/unlock-js": "workspace:^",
```
Comment (searchableguy) [inline packages/paywall/package.json:36]:
```suggestion

```

File: packages/paywall/package.json (hunk 7 above)
Comment (searchableguy) [reply packages/paywall/package.json:41]: Yes

File: unlock-app/src/hooks/usePostmateParent.ts [hunk 17]
```diff
@@ -1,6 +1,8 @@
 import { useState, useEffect } from 'react'
 import Postmate from 'postmate'
 
+Postmate.debug = true
```
Comment (searchableguy) [inline unlock-app/src/hooks/usePostmateParent.ts:4]:
```suggestion
Postmate.debug = true
```

**Review by julien51 — APPROVED** [review]
Lgtm

**Pull request merged**